patches/022-openssl-negative-tests.patch

```diff
@@ -0,0 +1,1003 @@
+index 3898f7f12c..e412a29269 100644
+--- a/src/crypto/hmac/hmac_test.go
++++ b/src/crypto/hmac/hmac_test.go
+@@ -13,6 +13,7 @@ import (
+ 	"crypto/sha512"
+ 	"fmt"
+ 	"hash"
++	"strings"
+ 	"testing"
+ )
+ 
+@@ -549,6 +550,16 @@ var hmacTests = []hmacTest{
+ 
+ func TestHMAC(t *testing.T) {
+ 	for i, tt := range hmacTests {
++		if boring.Enabled() && tt.size == sha1.New().Size() {
++			defer func() {
++				r := recover()
++				if s, ok := r.(string); ok {
++					if !strings.Contains(s, "unrecognized hmac in FIPS mode") {
++						panic(s)
++					}
++				}
++			}()
++		}
+ 		h := New(tt.hash, tt.key)
+ 		if s := h.Size(); s != tt.size {
+ 			t.Errorf("Size: got %v, want %v", s, tt.size)
+@@ -623,6 +634,9 @@ func TestEqual(t *testing.T) {
+ 
+ func TestHMACHash(t *testing.T) {
+ 	for i, test := range hmacTests {
++		if boring.Enabled() && test.size == sha1.New().Size() {
++			t.Skip("skipping test, negative assertion (failure mode) covered by TestHMAC")
++		}
+ 		baseHash := test.hash
+ 		key := test.key
+ 
+diff --git a/src/crypto/rsa/pkcs1v15_test.go b/src/crypto/rsa/pkcs1v15_test.go
+index 0e38bbfbef..a03aa1cb62 100644
+--- a/src/crypto/rsa/pkcs1v15_test.go
++++ b/src/crypto/rsa/pkcs1v15_test.go
+@@ -17,6 +17,7 @@ import (
+ 	"encoding/hex"
+ 	"encoding/pem"
+ 	"io"
++	"strings"
+ 	"testing"
+ 	"testing/quick"
+ )
+@@ -55,10 +56,6 @@ var decryptPKCS1v15Tests = []DecryptPKCS1v15Test{
+ }
+ 
+ func TestDecryptPKCS1v15(t *testing.T) {
+-	if boring.Enabled() && !boringtest.Supports(t, "PKCSv1.5") {
+-		t.Skip("skipping PKCS#1 v1.5 encryption test with BoringCrypto")
+-	}
+-
+ 	decryptionFuncs := []func([]byte) ([]byte, error){
+ 		func(ciphertext []byte) (plaintext []byte, err error) {
+ 			return DecryptPKCS1v15(nil, rsaPrivateKey, ciphertext)
+@@ -72,7 +69,14 @@ func TestDecryptPKCS1v15(t *testing.T) {
+ 		for i, test := range decryptPKCS1v15Tests {
+ 			out, err := decryptFunc(decodeBase64(test.in))
+ 			if err != nil {
+-				t.Errorf("#%d error decrypting: %v", i, err)
++				if boring.Enabled() && !boringtest.Supports(t, "PKCSv1.5") {
++					if !strings.Contains(err.Error(), "decryption error") {
++						t.Fatalf("unexpected error in FIPS mode: %s", err.Error())
++					}
++					continue
++				} else {
++					t.Errorf("#%d error decrypting: %v", i, err)
++				}
+ 			}
+ 			want := []byte(test.out)
+ 			if !bytes.Equal(out, want) {
+@@ -83,10 +87,6 @@ func TestDecryptPKCS1v15(t *testing.T) {
+ }
+ 
+ func TestEncryptPKCS1v15(t *testing.T) {
+-	if boring.Enabled() && !boringtest.Supports(t, "PKCSv1.5") {
+-		t.Skip("skipping PKCS#1 v1.5 encryption test with BoringCrypto")
+-	}
+-
+ 	random := rand.Reader
+ 	k := (rsaPrivateKey.N.BitLen() + 7) / 8
+ 
+@@ -97,6 +97,11 @@ func TestEncryptPKCS1v15(t *testing.T) {
+ 
+ 		ciphertext, err := EncryptPKCS1v15(random, &rsaPrivateKey.PublicKey, in)
+ 		if err != nil {
++			if boring.Enabled() && !boringtest.Supports(t, "PKCSv1.5") {
++				if strings.Contains(err.Error(), "invalid key length") {
```

Comment:
Can we use a larger key instead of `rsaPrivateKey` to remove this check? Also, do we want to actually check the key size, maybe with a separate helper function?

Reply:
Yeah, I think that checking the key size is a good idea, but I don't think we should *change* the size in the tests, just to ensure we can validate we get this error when we expect to.